Add remote file handling

b-rowan opened the pull request:
Adds remote firmware handling for files.  Based on #29.  Closes #28. Closes #29.

Adding independently so they can be merged independently if there are concerns on this PR.

Main concern here is that I don't have a better way to get the hash of the upgrade file (required for swupdate artifact) without just downloading the file when the link is added.

easybe commented on goosebit/models.py:
```diff
@@ -38,3 +45,65 @@ class Rollout(Model):
+    hw_revision = fields.CharField(max_length=255)
+
+
+class FirmwareUpdate(Model):
```
I would just call it `Firmware` as it does not have to be an update, it could technically also be a downgrade (if we would start to allow that) or the initial image.

tsagadar replied:
The rollout model needs to be adjusted as well - instead of hw_model, hw_revision, fw_file it only should reference a firmware.

This then also means work in Manager.get_rollout to query the correct firmware.

b-rowan replied:
> The rollout model needs to be adjusted as well - instead of hw_model, hw_revision, fw_file it only should reference a firmware.
> 
> This then also means work in Manager.get_rollout to query the correct firmware.

This can get really tricky, if I link `Firmware` directly then all the stuff that wants to check for rollout that uses the device as a backup need to return `Firmware`, which means `fw_file` on `Device` needs to be replaced with `Firmware`, but then that gets rid of `latest` and `pinned`, which can make things pretty complicated.  Let me know what you guys think here.

tsagadar replied:
Yes, it triggers a few follow-up changes. I have a branch where I introduced `update_mode` on Device (enum with `latest`, `rollout`, `assigned`, and `pinned`), so that `fw_file` could be changed to `firmware` on Device.

Will re-implement those changes on top of this branch to see if it makes sense.

easybe reviewed (commented):
IMHO downloading the artifact when creating a new DB entry to get the hashes is a reasonable approach.

I’ve only skimmed over the changes as reviewing on the phone is not that fun.

easybe commented:
Oh, and please note this comment: https://github.com/UpstreamDataInc/goosebit/commit/c5531765c7e98009e76a81d4d656c04370653561#r144602289

tsagadar commented:
I think we should slow down somewhat: as there are no automatic test in the code base, every PR breaks some functionality. Manual testing (what I was doing so far) takes forever. So we better ensure to have tests for the existing functionality before adding / refactoring code.

I created automated API integration tests (focus on the DDI) yesterday - and currently adjust the test code to match this PR. Ideally I could finish these tests (few of them will fail as the code is broken), integrate them into this PR and ensure that everything passes before merging.

tsagadar commented on goosebit/models.py:
```diff
@@ -38,3 +45,65 @@ class Rollout(Model):
+    )
+
+    @classmethod
+    async def latest(cls):
```
This needs a parameter "device" to properly filter for the latest firmware.

tsagadar commented on goosebit/updater/manager.py:
```diff
@@ -170,16 +169,26 @@ async def get_update_file(self) -> FirmwareArtifact:
         device = await self.get_device()
 
         file = await self.get_update_file()
-        if file.is_empty():
```
Why removing this case? At least when the system starts and devices connect there are no firmware available and no update should be rolled out.

tsagadar commented on goosebit/updater/manager.py:
```diff
@@ -170,16 +169,26 @@ async def get_update_file(self) -> FirmwareArtifact:
     async def get_update_mode(self) -> str:
         device = await self.get_device()
 
         file = await self.get_update_file()
```
Rename file / get_update_file to firmware / get_firmware (or similar)

tsagadar commented on goosebit/updater/manager.py:
```diff
@@ -170,16 +169,26 @@ async def get_update_file(self) -> FirmwareArtifact:
-            mode = "skip"
-            self.poll_time = POLL_TIME
-        elif file.name == device.fw_version and not self.force_update:
+        if file.path.name.split(".")[0] == device.fw_version and not self.force_update:
```
Version is explicitly stored on firmware, so no need parsing the filename.

tsagadar commented on .pre-commit-config.yaml:
```diff
@@ -14,12 +14,12 @@ repos:
-        pass_filenames: false
-        always_run: true
\ No newline at end of file
+#-   repo: local
```
Why remove this in the first place? Should really have tests in the system.

tsagadar commented on goosebit/api/download.py:
```diff
@@ -1,20 +1,29 @@
-async def download_file(request: Request, file: str):
-    filename = UPDATES_DIR.joinpath(file)
-    return FileResponse(filename, media_type="application/octet-stream")
+@router.get("/by_name/{filename}")
```
What will this endpoint be used for?

b-rowan replied:
Fair enough, I think this can safely be removed, as all firmware can now just be queried by ID, even in the `/firmware/all` endpoint it reports the ID.

tsagadar commented on goosebit/db.py:
```diff
@@ -22,6 +23,10 @@ async def init():
     await command.migrate()
     await command.upgrade(run_in_transaction=True)
     await Tortoise.generate_schemas(safe=True)
+    for update in await FirmwareUpdate.all():
```
Instead of `all()` better to filter on db using `filter()`.

b-rowan replied:
Not sure what to filter by here, I suppose maybe `uri` starts with `file`?

tsagadar replied:
Thought about using `await Firmware.filter(url__startswith='file').all()`. Debatable if a good idea, as logic of `local` is the implemented in two different places. On the upside: with a lot of remote firmware it will be very fast.

tsagadar commented on goosebit/models.py:
```diff
@@ -38,3 +45,65 @@ class Rollout(Model):
+    def local(self):
+        return urlparse(self.uri).scheme == "file"
+
+    def generate_chunk(self, request: Request) -> list:
```
Starting to overload the model classes? Looks like an API related aspect.

b-rowan replied:
Moved to take `FirmwareUpdate` as a parameter.

tsagadar reviewed (changes requested):
Few things I saw while writing tests. Some if them cause my current tests to fail.

tsagadar reviewed (commented):
I finished a first set of integration test - they can be found at https://github.com/husqvarnagroup/goosebit/tree/mm/add_simple_ddi_integration_tests

Besides the commit adding the tests there is an additional one that makes the tests pass. Most of the things are also annotated in this PR.

Note: `poetry run pytest` does not pass all the tests. Single execution of tests work (I do it using PyCharm). Will look into this soon (but not today)

b-rowan commented:
I'll try to take a look at this later tonight, at a conference this week so I don't have a ton of time, but I'll fix what I can ASAP.

> I think we should slow down somewhat: as there are no automatic test in the code base, every PR breaks some functionality. 

Agreed, but I really want to get the important core functionality we want figured out ASAP.  I'm not a big tests guy personally, I think it's definitely a good idea in this case, I just struggle with the differences between trying to finish important functionality and stabilizing the code/API.  There seem to be a lot of cases where the dev API needs to change for important things to happen, so I wasn't sure if we wanted to wait to write tests until after that.

b-rowan commented on goosebit/updates/swdesc.py:
```diff
@@ -0,0 +1,52 @@
+        swdesc_attrs["version"] = semver.Version.parse(swdesc["software"]["version"])
+        swdesc_attrs["size"] = file.stat().st_size
+        swdesc_attrs["hash"] = sha1_hash_file(file)
+        swdesc_attrs["compatibility"] = [
```
Is there a standard on how this is supposed to be set?  We use `loadsync.v1` here, but I'm not sure if there is any standard here...

tsagadar replied:
This is not my area of expertise. Maybe @easybe can provide information once he is back.

b-rowan commented on goosebit/updater/manager.py:
```diff
@@ -170,16 +169,26 @@ async def get_update_file(self) -> FirmwareArtifact:
+            return await FirmwareUpdate.latest()
+        if file == "pinned":
+            return await FirmwareUpdate.get(
+                uri=UPDATES_DIR.joinpath(device.fw_file).absolute().as_uri()
```
Need to take a look at this, not sure how the best way to have it still generate the artifact information and have it just skip with no file.  All the artifact generation being done in the firmware file makes things a bit tough, maybe I should split that method out and have it take `Firmware | None`?

b-rowan commented:
> I finished a first set of integration test - they can be found at https://github.com/husqvarnagroup/goosebit/tree/mm/add_simple_ddi_integration_tests
> 
> Besides the commit adding the tests there is an additional one that makes the tests pass. Most of the things are also annotated in this PR.
> 
> Note: `poetry run pytest` does not pass all the tests. Single execution of tests work (I do it using PyCharm). Will look into this soon (but not today)

Implemented these, and fixed your issue with pytest not working properly.  Just need to add `__init__.py` files to all directories leading to the tests to make them identify as modules.

b-rowan commented:
Mostly everything fixed for now, just a couple things I want to clarify on before I change anything else, check comments on unresolved when you get a chance @tsagadar @easybe.  Appreciate the pointers, this was a pretty large PR, but I think a pretty important one to deal with early on.

tsagadar commented:
I think this PR is ready to be merged - great stuff. Follow-up work can be done in a new branch / PR. Will look replacing fw_file with firmware.
